### PR TITLE
Implemented MD5 based deployment

### PR DESCRIPTION
**Pull request by sebjulliand**
🎅🎄Ho ho ho, Merry Christmas! 🎄🎅
### Changes
Added a new deployment method called `MD5 compare`. It uses files' MD5 hash from local workspace and remote deployment location to:
* Upload only changed files since last deployment
* Delete remote files not found in the local workspace

It also clears empty folders in the remote deployment location.

The feature uses the `md5sum` shell command on the remote system. The extension will check if the command is available in order to enable the feature.

The `deploy` method has also been added to the exported function in the extension's API, to let other extension run unattended deployments.

### Checklist
* [x] have tested my change
* [x] updated relevant documentation
* [x] **for feature PRs**: PR only includes one feature enhancement.

### CODE REVIEW

File: src/api/local/deployment.ts [hunk 1]
```diff
@@ -172,13 +176,18 @@ export namespace Deployment {
...
+          { method: Method.unstaged, label: `Working Changes`, description: `Unstaged changes in git` },
+          { method: Method.staged, label: `Staged Changes`, description: `` },
+          { method: Method.all, label: `All`, description: `Every file in the local workspace` }
+        ];
+
+        if (getConnection().remoteFeatures.md5sum) {
+          methods.push({ method: Method.md5synch, label: `MD5 compare`, description: `Synchronizes using MD5 hash comparison` });
+        }
+
+        const method = (await vscode.window.showQuickPick(methods.sort((m1, m2) => m1.label.localeCompare(m2.label)),
```
Comment (worksofliam) [inline src/api/local/deployment.ts:190]: I am not sure how I feel about this sort.

This means `All` is always showing up at the top, and it's rare that we would want 'all' every time. I think we should aim to have `Changed` or `MD5 Compare` at the top.

File: src/api/local/deployment.ts [hunk 2]
```diff
@@ -172,13 +176,18 @@ export namespace Deployment {
...
-          ],
+        const methods = [
+          { method: Method.changed, label: `Changes`, description: `${changes.size} change${changes.size > 1 ? `s` : ``} detected since last upload. ${!changes.size ? `Will skip deploy step.` : ``}` },
+          { method: Method.unstaged, label: `Working Changes`, description: `Unstaged changes in git` },
+          { method: Method.staged, label: `Staged Changes`, description: `` },
+          { method: Method.all, label: `All`, description: `Every file in the local workspace` }
+        ];
+
+        if (getConnection().remoteFeatures.md5sum) {
+          methods.push({ method: Method.md5synch, label: `MD5 compare`, description: `Synchronizes using MD5 hash comparison` });
```
Comment (worksofliam) [inline src/api/local/deployment.ts:187]: Personally, I think we need a different name than 'MD5 Compare'. Perhaps even just `Compare` is ok, as the description has got it covered.

**Review by worksofliam — CHANGES_REQUESTED** [review]
This is very awesome change. I've tested in with a few different situations and it is looking good. My only concern right now is that I don't want the 'md5 compare' upload method to upload items defined in my `.gitignore`. 

It doesn't look like the `deployMD5` is looking at the ignore rules. I think if a file in the `uploads` is in the ignore rules, then it shouldn't be uploaded.

What do you think?

File: src/api/local/deployment.ts (hunk 1 above)
Comment (sebjulliand) [reply src/api/local/deployment.ts:190]: Sounds good to me! I'll put both on top (`Compare` then `Changed`).

File: src/api/local/deployment.ts (hunk 2 above)
Comment (sebjulliand) [reply src/api/local/deployment.ts:187]: Let's go with `Compare` then. 👍

**Comment by sebjulliand** [comment]
> This is very awesome change. I've tested in with a few different situations and it is looking good. My only concern right now is that I don't want the 'md5 compare' upload method to upload items defined in my `.gitignore`.
> 
> It doesn't look like the `deployMD5` is looking at the ignore rules. I think if a file in the `uploads` is in the ignore rules, then it shouldn't be uploaded.
> 
> What do you think?

You're right to ask, and actually, `Compare` does not uploads what's in `.gitignore`. At least, it's not supposed to. The exclusion is done in the `findFiles` method which lists all the local files but those matching a pattern from `.gitignore`.

File: src/api/local/deployment.ts [hunk 3]
```diff
@@ -474,4 +578,39 @@ export namespace Deployment {
...
+  function toRelative(root: vscode.Uri, file: vscode.Uri) {
+    return path.relative(root.path, file.path).replace(/\\/g, `/`);
+  }
+
+  async function findFiles(parameters: DeploymentParameters, includePattern: string, excludePattern?: string) {
+    const root = parameters.localFolder.fsPath;
+    return (await vscode.workspace.findFiles(new vscode.RelativePattern(root, includePattern),
+      excludePattern ? new vscode.RelativePattern(root, excludePattern) : null))
+      .filter(file => {
+        if (parameters.ignoreRules) {
```
Comment (sebjulliand) [inline src/api/local/deployment.ts:607]: @worksofliam: this is where the files matching `.gitignore` get excluded for the `Compare` method.

Comment (worksofliam) [reply src/api/local/deployment.ts:607]: Thanks for pointing that out for me!

**Comment by worksofliam** [comment]
Review in coming later today.

**Comment by worksofliam** [comment]
LGTM. Thank you @sebjulliand - this is really an outstanding change.

**Review by worksofliam — APPROVED** [review]

**Pull request merged**